pkg/util/certificates.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+)
+
+func getCACertPool(caFile string) (*x509.CertPool, error) {
+	certPool := x509.NewCertPool()
+	caCert, err := ioutil.ReadFile(caFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read agent CA cert %s: %v", caFile, err)
+	}
+	ok := certPool.AppendCertsFromPEM(caCert)
+	if !ok {
+		return nil, fmt.Errorf("failed to append CA cert to the cert pool")
+	}
+	return certPool, nil
+}
+
+// loadCertificateFromAPIServer loads certificate from kube-apiserver by establishing tls connection and pulling cert from it
+func loadCertificateFromAPIServer(caFile, apiHost string) (*x509.Certificate, error) {
+	rootCas, err := getCACertPool(caFile)
+	if err != nil {
+		return nil, err
+	}
+	conf := &tls.Config{
+		RootCAs: rootCas,
+	}
+
+	apiAddr := apiHost + ":443"
+	conn, err := tls.Dial("tcp", apiAddr, conf)
+	if err != nil {
+		return nil, fmt.Errorf("failed dial to:%q. err:%v", apiAddr, err)
+	}
+	defer conn.Close()
+
+	if len(conn.ConnectionState().PeerCertificates) < 1 {
+		return nil, fmt.Errorf("no certificates was pulled from %q", apiHost)
+	}
+
+	return conn.ConnectionState().PeerCertificates[0], nil
+}
+
+// GetTLSConfigwithAPIAuth returns tlsConfig based on kube-apiserver auth
+func GetTLSConfigwithAPIAuth(caFile, apiHost string) (*tls.Config, error) {
+	c, err := loadCertificateFromAPIServer(caFile, apiHost)
+	if err != nil {
+		return nil, err
+	}
+
+	certPool := x509.NewCertPool()
+	certPool.AddCert(c)
+
+	return &tls.Config{RootCAs: certPool}, nil
```

Comment:
@mikedanese @dberkov based on the offline discussion, I think we just need to put here the CA cert used by kubelet, then the proxy agent can validate the proxy server.

Reply:
Done.